src/converters/intrinsics.ts

```diff
@@ -0,0 +1,295 @@
+// Copyright 2016-2024, Pulumi Corporation.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import * as equal from 'fast-deep-equal';
+
+/**
+ * Models a CF Intrinsic Function.
+ *
+ * CloudFormation (CF) intrinsic functions need to be implemented for @pulumi/pulumi-cdk since CDK may emit them in the
+ * synthesized CF template.
+ *
+ * See https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/intrinsic-function-reference.html
+ *
+ * @internal
+ */
+export interface Intrinsic {
+    /**
+     * The name of the intrinsic function such as 'Fn::If'.
+     */
+    name: string;
+
+    /**
+     * Executes the logic to evaluate CF expressions and compute the result.
+     *
+     * Most intrinsics need to use IntrinsicContext.evaluate right away to find the values of parameters before
+     * processing them. Conditional intrinsics such as 'Fn::If' or 'Fn::Or' are an exception to this and need to
+     * evaluate their parameters only when necessary.
+     */
+    evaluate(ctx: IntrinsicContext, params: Expression[]): Result<any>;
+}
+
+/**
+ * Models a CF expression. Currently this is just 'any' but eventually adding more structure and a separate
+ * parse/evaluate steps can help keeping the error messages tractable.
+ *
+ * See also CfnParse for inspiration:
+ *
+ * https://github.com/aws/aws-cdk/blob/main/packages/aws-cdk-lib/core/lib/helpers-internal/cfn-parse.ts#L347
+ *
+ * @internal
+ */
+export interface Expression {}
+
+/**
+ * Production code may have intermediate values occasionally wrapped in pulumi.Output<T>; this is currently somewhat
+ * difficult to test, so the essentials of pulumi.Output<T> are abstracted into a Result<T>.
```

Comment:
I don't quite understand this. Right now it's just `any`, isn't it?

Reply:
It's `pulumi.Output<T>` in prod, but `TestResult<T>` in test. I could have used `pulumi.Output<T>` but currently testing failing scenarios there is very annoying.